### ISSUE TITLE
diff-so-fancy slurps entire git log -p output?  very slow

### ISSUE THREAD

**Issue by bronson** [so-fancy/diff-so-fancy#140]
Using Feb 7th's diff-so-fancy, `git log -p` and `git log --stat` would display immediately.

Using today's diff-so-fancy, `git log --stat` takes many seconds to complete and `git log -p` hangs for a few minutes and then bombs out.

A guess: it looks like some change in the past 1.5 months made it so diff-so-fancy is now trying to slurp the entire output before displaying the results.

My repo has 4200 commits, so tiny.  This issue appears reproducible in any moderately-sized project.  For example, `git log --stat` on rails/rails shows a blank screen for 10 seconds before showing the log.  Revert to Feb 7th and it displays immediately.

**Comment by stevemao** [comment]
I don't think you should use `diff-so-fancy` for log pager.

As per readme:
```
git config --global pager.diff "diff-so-fancy | less --tabs=4 -RFX"
git config --global pager.show "diff-so-fancy | less --tabs=4 -RFX"
```

**Comment by scottchiefbaker** [comment]
I agree with @stevemao. d-s-f is designed to process small diffs (commits) in to human readable form, not to parse hundreds of thousands of lines. 

My guess for the slowness is that we remove all the leading `+` and `-` from the diff before we process any actual content. It'd be pretty hard to work around that for this one use case.

**Comment by bronson** [comment]
Ah, the recommended config changed on Feb 12: 5f7422aa38c0bd16640f85dc2f02b72f85e8af88

Updated my config to match and all is fine.  Thanks.

**Comment by bronson** [comment]
That said....   it's interesting that Feb 7th's d-s-f could handle being the core.pager and today's can't.

**Comment by blueyed** [comment]
I agree that it would be better if it still could be used with `log`.

It would be helpful to find the commit that caused this (using git-bisect) and then improve upon that.
Even if the removal/processing of all +/- would cause this, this could still be done in a pipeline then probably?!

**Comment by scottchiefbaker** [comment]
I think @bronson found it at a6b5801. That's the commit where we switched most of the backend from sed to Perl so we could do fancier things with the output display. We can investigate the slowness, but I can't make any guarantees as this use case is a little outside the scope of the project.

**Comment by blueyed** [comment]
It should still be possible to not read the whole input, but process it in chunks (e.g. 8kb).

**Comment by scottchiefbaker** [comment]
If I use a 60MB `git log -p` output I see the following:

* 1 second to load the whole input in to memory
* 7 seconds to clean up the input before we process anything

The rest of the time is spent looping through each line one by one... I imagine the problem is the second one (depending on how much input we're talking). As a work around you can completely disable the line pre-processing by commenting out line 22

https://github.com/so-fancy/diff-so-fancy/blob/master/lib/diff-so-fancy.pl#L22

That will get you output at the 1 second mark.

**Comment by blueyed** [comment]
@scottchiefbaker 
Thanks for investigating.

Looks like the `clean_up_input` should be moved inside of the loop?!

**Comment by bronson** [comment]
@blueyed agreed, the problem appears to be that this script slurps the entire file before processing.  Moving the `<>` into the loop seems like it would bring the old behavior back.

```perl
while(my $line = <>) {
    ...
}
```

But it seems like this script goes out of its way to slurp everything before processing...  not sure if there's a reason for that.

**Comment by scottchiefbaker** [comment]
This actually ended up not being too insane to solve, and it will give us other performance improvements. Can you test the branch I just posted here:

https://github.com/scottchiefbaker/diff-so-fancy/tree/pipeline

If I use my 60MB input now, I get output on the screen immediately. I moved back tp processing the input line by line instead of buffering the whole input. This is a little more complex code wise, but it should perform a lot better.

**Comment by scottchiefbaker** [comment]
Also, most importantly, we still pass all tests.

**Comment by bronson** [comment]
@scottchiefbaker looks great!

It appears nobody uses `$i` anymore...  seems like you can drop those lines too?

And, if anyone needs a line number in the future, maybe they could use `$.`.

**Comment by scottchiefbaker** [comment]
True... `$i` is used to check if it's the first line. I was going to use it other places, but ended up not needing it. It's fine for now. I'll reopen this since we have a fix and wait for feedback on whether it does what you want or not.

**Comment by bronson** [comment]
@scottchiefbaker just tried it out, everything appears to work perfect.  Even `git log -p` in rails/rails.

**Comment by scottchiefbaker** [comment]
How big is rails/rails?

**Comment by bronson** [comment]
Dunno, doesn't matter!  :)  Now d-s-f blocks before getting very far.

Over 57,000 commits, gotta be hundreds of MB if one lets it run to
completion.

**Comment by scottchiefbaker** [comment]
@stevemao @paulirish what do you guys think? Should we merge this even though it's outside the scope of regular d-s-f?

**Comment by stevemao** [comment]
It's not out of the scope. being smart on processing the lines will give performance boost on `git diff` too.

**Comment by blueyed** [comment]
Thanks, @scottchiefbaker!
Please create a PR for it.

**Comment by scottchiefbaker** [comment]
This is merged so I'm closing this issue now.

**Comment by bronson** [comment]
Love it, great job!

**Comment by Blaisorblade** [comment]
Any chance this problem slipped in again? `git log -p` takes lots (on `commercialhaskell/stack`) on `git log -p` when enabled as default pager. Lots seems to be 10-20s (timed it to ~15s) rather than instantaneous, so it took me a few minutes of debugging (and looking at this issue) to confirm I just had to wait for output to appear.

**Comment by scottchiefbaker** [comment]
@Blaisorblade this should be fixed on the `next` branch. It will be released officially as 1.2.0 when we're ready. See #251 for more info.

**Comment by bryantee** [comment]
Thanks for clearing up @scottchiefbaker
